ext/rubymain.cpp

```diff
@@ -111,6 +115,155 @@ static inline VALUE ensure_conn(const uintptr_t signature)
 	return conn;
 }
 
+/****************
+em_ssl_verify_cb_call
+****************/
+
+#ifdef WITH_SSL
+
+// adapted from stdlib openssl's ossl_str_new_i
+static VALUE em_ssl_str_new_i(VALUE size)
+{
+	return rb_str_new(NULL, (long)size);
+}
+
+// adapted from stdlib openssl's ossl_str_new
```

Comment:
Would it be possible to use those method directly (if they are part of the public API?)

Reply:
For the most part, yes, and I would love to just use the methods directly.  Even better, I'd love to just use `OpenSSL::X509::StoreContext` directly, and skip all of the re-implementation.

The problem (as I understand it) is that there's no mechanism for a ruby gem to publish C header files that can safely be used (not even for bundled or default gems, like `openssl`).  And even then, we'd also need the `openssl` gem to commit to a public API for any C funcs that we used... many of the functions I want access to are static and not exported.  There's at least one b.r-l.org conversation about this issue: https://bugs.ruby-lang.org/issues/17760, but it didn't really come to any resolution, AFAICT.

That said, it _should_ be possible to find the gem's install dir and use ext/openssl/ossl.h.  And the most important functions that we might want to access _are_ exported.  I just don't have any idea how brittle that would be... and _when_ it breaks, how hard to debug will it be?

Reply:
What you've done makes sense given that these are internal methods that aren't being _intentionally_ exported with a stable API. Someone put the idea in my head ages ago to move all of the TLS work up into Ruby space and implement it using the ruby/openssl standard library gem, and just rip out all of the C++ TLS bits. That's not something I've had any time in years to think much about, but would seem to be the right move.

Reply:
Adding a few more links I found:
* https://bugs.ruby-lang.org/issues/17760
* https://bugs.ruby-lang.org/issues/17761 (Fixed by: https://github.com/ruby/ruby/commit/5cdf99f6)
* https://bugs.ruby-lang.org/issues/15156
* https://github.com/rubygems/rubygems/issues/2408
  * https://github.com/rubygems/rubygems/pull/2416
* https://github.com/ruby/openssl/issues/475
  _I created this issue on the openssl gem back when I first worked on this current eventmachine PR._